File: internal/handler/mcp_gateway.go
package handler

import (
	"net/http"

	"github.com/Tencent/WeKnora/internal/application/service"
	"github.com/Tencent/WeKnora/internal/errors"
	"github.com/Tencent/WeKnora/internal/logger"
	"github.com/Tencent/WeKnora/internal/types"
	secutils "github.com/Tencent/WeKnora/internal/utils"
	"github.com/gin-gonic/gin"
)

// MCPGatewayHandler handles MCP → agent-gateway synchronization requests.
//
// These endpoints do NOT touch the /mcp-services CRUD: WeKnora stays the source
// of truth for MCP config, and this handler only pushes a derived copy to the
// agent-gateway (the process that actually executes agent runs).
type MCPGatewayHandler struct {
	mcpGatewaySyncService *service.MCPGatewaySyncService
}

// NewMCPGatewayHandler creates a new MCP gateway sync handler
func NewMCPGatewayHandler(mcpGatewaySyncService *service.MCPGatewaySyncService) *MCPGatewayHandler {
	return &MCPGatewayHandler{
		mcpGatewaySyncService: mcpGatewaySyncService,
	}
}

// SyncMCPGateway godoc
// @Summary      同步MCP服务到Agent网关
// @Description  把当前空间的MCP服务配置全量覆盖下发到agent-gateway（幂等）
// @Tags         MCP服务
// @Accept       json
// @Produce      json
// @Success      200  {object}  map[string]interface{}  "同步结果（synced/skipped/warnings）"
// @Failure      400  {object}  errors.AppError         "请求参数错误"
// @Failure      500  {object}  errors.AppError         "服务器错误"
// @Security     Bearer
// @Security     ApiKeyAuth
// @Router       /mcp-gateway/sync [post]
func (h *MCPGatewayHandler) SyncMCPGateway(c *gin.Context) {
	ctx := c.Request.Context()

	tenantID := c.GetUint64(types.TenantIDContextKey.String())
	if tenantID == 0 {
		logger.Error(ctx, "Tenant ID is empty")
		c.Error(errors.NewBadRequestError("Workspace ID cannot be empty"))
		return
	}

	result, err := h.mcpGatewaySyncService.Sync(ctx, tenantID)
	if err != nil {
		logger.ErrorWithFields(ctx, err, map[string]interface{}{"tenant_id": tenantID})
		c.Error(errors.NewInternalServerError("Failed to sync MCP services to agent-gateway: " + err.Error()))
		return
	}

	logger.Infof(ctx, "MCP services synced to agent-gateway: %d synced, %d skipped, %d warnings",
		len(result.Synced), len(result.Skipped), len(result.Warnings))
	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    result,
	})
}

// MCPGatewayStatus godoc
// @Summary      查询Agent网关的MCP配置状态
// @Description  对比WeKnora与agent-gateway的MCP配置差异（drift检测）
// @Tags         MCP服务
// @Accept       json
// @Produce      json
// @Success      200  {object}  map[string]interface{}  "网关状态"
// @Failure      400  {object}  errors.AppError         "请求参数错误"
// @Failure      500  {object}  errors.AppError         "服务器错误"
// @Security     Bearer
// @Security     ApiKeyAuth
// @Router       /mcp-gateway/status [get]
func (h *MCPGatewayHandler) MCPGatewayStatus(c *gin.Context) {
	ctx := c.Request.Context()

	tenantID := c.GetUint64(types.TenantIDContextKey.String())
	if tenantID == 0 {
		logger.Error(ctx, "Tenant ID is empty")
		c.Error(errors.NewBadRequestError("Workspace ID cannot be empty"))
		return
	}

	status, err := h.mcpGatewaySyncService.Status(ctx, tenantID)
	if err != nil {
		logger.ErrorWithFields(ctx, err, map[string]interface{}{"tenant_id": tenantID})
		c.Error(errors.NewInternalServerError("Failed to get MCP gateway status: " + err.Error()))
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    status,
	})
}

// TestMCPGatewayServer godoc
// @Summary      通过Agent网关测试MCP服务
// @Description  以agent-gateway的视角测试MCP服务连通性并回显工具列表
// @Tags         MCP服务
// @Accept       json
// @Produce      json
// @Param        id   path      string  true  "MCP服务ID"
// @Success      200  {object}  map[string]interface{}  "网关侧测试结果"
// @Failure      400  {object}  errors.AppError         "请求参数错误"
// @Security     Bearer
// @Security     ApiKeyAuth
// @Router       /mcp-gateway/servers/{id}/test [post]
func (h *MCPGatewayHandler) TestMCPGatewayServer(c *gin.Context) {
	ctx := c.Request.Context()
	serviceID := secutils.SanitizeForLog(c.Param("id"))

	tenantID := c.GetUint64(types.TenantIDContextKey.String())
	if tenantID == 0 {
		logger.Error(ctx, "Tenant ID is empty")
		c.Error(errors.NewBadRequestError("Workspace ID cannot be empty"))
		return
	}

	logger.Infof(ctx, "Testing MCP service through agent-gateway: %s", secutils.SanitizeForLog(serviceID))

	result, err := h.mcpGatewaySyncService.TestServer(ctx, tenantID, serviceID)
	if err != nil {
		// Mirrors TestMCPService: a failed probe is reported as data with
		// ok=false rather than an error envelope, so the UI can render the
		// reason inline. The gateway itself also answers 200 + ok=false for
		// unreachable servers, so the two shapes stay identical.
		logger.ErrorWithFields(ctx, err, map[string]interface{}{"service_id": secutils.SanitizeForLog(serviceID)})
		c.JSON(http.StatusOK, gin.H{
			"success": true,
			"data": map[string]interface{}{
				"ok":    false,
				"error": err.Error(),
			},
		})
		return
	}

	logger.Infof(ctx, "MCP service gateway test completed: %s", secutils.SanitizeForLog(serviceID))
	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    result,
	})
}

File: frontend/src/api/mcp-gateway.ts
import { get, post } from '@/utils/request'

// ---------------------------------------------------------------------------
// MCP → agent-gateway 同步（私有化定制）
//
// WeKnora 是 MCP 配置的唯一事实源（source of truth），这里只把一份派生副本
// 全量覆盖下发到真正执行 agent 的 agent-gateway。
//
// 后端契约（internal/handler/mcp_gateway.go + internal/application/service/mcp_gateway_sync.go）：
//   POST /api/v1/mcp-gateway/sync                  → {success:true, data:MCPGatewaySyncResult}
//   GET  /api/v1/mcp-gateway/status                 → {success:true, data:MCPGatewayStatusResult}
//   POST /api/v1/mcp-gateway/servers/:id/test       → {success:true, data:MCPGatewayTestResult}
//
// 响应取值方式与 api/mcp-service.ts 保持一致：utils/request.ts 的响应拦截器
// 已经把整个 HTTP body 作为 resolve 值返回，所以这里拿到的是
// { success, data, msg }；统一用 response.data ?? response 读出业务数据。
// ---------------------------------------------------------------------------

/** 网关侧跳过 / 警告的单项说明 */
export interface MCPGatewaySkipItem {
  name: string
  reason: string
}

/** POST /api/v1/mcp-gateway/sync 的 data */
export interface MCPGatewaySyncResult {
  gateway_url: string
  /** 已成功下发到网关的服务名 */
  synced: string[]
  /** 被跳过的服务及原因（未启用 / SSE 不支持 / 配置缺失等） */
  skipped: MCPGatewaySkipItem[]
  /** 下发成功但需注意的服务（如 OAuth 无法复用 token） */
  warnings: MCPGatewaySkipItem[]
  /** 网关侧配置文件来源：managed（托管文件）| seed（种子文件） */
  source: string
  /** 本次下发时间 */
  pushed_at: string
}

/** GET /api/v1/mcp-gateway/status 的 data */
export interface MCPGatewayStatusResult {
  gateway_url: string
  reachable: boolean
  /** 网关侧配置文件来源：managed（托管文件）| seed（种子文件） */
  source: string
  /** 网关当前生效的服务数量 */
  count: number
  /** 网关侧的服务名列表 */
  gateway_names: string[]
  /** WeKnora 已启用但网关缺少（drift） */
  missing_on_gateway: string[]
  /** 网关多出、WeKnora 侧已删除或停用（drift） */
  extra_on_gateway: string[]
  /** 网关不可达等原因 */
  error?: string
}

/** 网关视角的工具描述 */
export interface MCPGatewayToolInfo {
  name: string
  description?: string
  /** 入参名列表（网关侧只回显参数名，不回显 schema） */
  params?: string[]
}

/** POST /api/v1/mcp-gateway/servers/:id/test 的 data */
export interface MCPGatewayTestResult {
  ok: boolean
  name?: string
  /** 网关侧传输类型，如 streamable_http / stdio */
  transport?: string
  tool_count?: number
  tools?: MCPGatewayToolInfo[]
  error?: string | null
  elapsed_ms?: number
}

/**
 * 全量下发当前空间的 MCP 服务到 agent-gateway（幂等、覆盖式）。
 * 仅在网关不可达 / 配置非法时抛错；被跳过的服务通过 skipped/warnings 返回。
 * force=true 才允许下发空配置（否则后端会以 400 拦截，避免误清空网关配置）。
 */
export async function syncMCPGateway(force = false): Promise<MCPGatewaySyncResult> {
  const response: any = await post(`/api/v1/mcp-gateway/sync${force ? '?force=true' : ''}`, {})
  return (response?.data ?? response) as MCPGatewaySyncResult
}

/** 查询 agent-gateway 当前生效的 MCP 配置，并与 WeKnora 侧对比出 drift。 */
export async function getMCPGatewayStatus(): Promise<MCPGatewayStatusResult> {
  const response: any = await get('/api/v1/mcp-gateway/status')
  return (response?.data ?? response) as MCPGatewayStatusResult
}

/**
 * 以 agent-gateway 的视角测试某个已配置服务的连通性，并回显网关侧真实工具列表。
 * 探测失败也返回 200 + ok=false（error 里带原因），不抛错。
 */
export async function testMCPGatewayServer(serviceId: string): Promise<MCPGatewayTestResult> {
  const response: any = await post(`/api/v1/mcp-gateway/servers/${encodeURIComponent(serviceId)}/test`, {})
  return (response?.data ?? response) as MCPGatewayTestResult
}
